internal/consensus/reactor.go

```diff
@@ -421,31 +416,58 @@ func (conR *Reactor) subscribeToBroadcastEvents() {
 	const subscriber = "consensus-reactor"
 	if err := conR.conS.evsw.AddListenerForEvent(subscriber, types.EventNewRoundStep,
 		func(data cmtevents.EventData) {
-			conR.broadcastNewRoundStepMessage(data.(*cstypes.RoundState))
-			conR.updateRoundStateNoCsLock()
+			rs := data.(cstypes.RoundState)
+
+			// update reactor's view of round state
+			conR.rsMtx.Lock()
+			conR.rs = rs
+			conR.rsMtx.Unlock()
+
+			conR.broadcastNewRoundStepMessage(&rs)
 		}); err != nil {
 		conR.Logger.Error("Error adding listener for events (NewRoundStep)", "err", err)
 	}
 
 	if err := conR.conS.evsw.AddListenerForEvent(subscriber, types.EventValidBlock,
 		func(data cmtevents.EventData) {
-			conR.broadcastNewValidBlockMessage(data.(*cstypes.RoundState))
+			rs := data.(cstypes.RoundState)
+
+			// update reactor's view of round state
+			conR.rsMtx.Lock()
+			conR.rs = rs
+			conR.rsMtx.Unlock()
+
+			conR.broadcastNewValidBlockMessage(&rs)
 		}); err != nil {
 		conR.Logger.Error("Error adding listener for events (ValidBlock)", "err", err)
 	}
 
 	if err := conR.conS.evsw.AddListenerForEvent(subscriber, types.EventVote,
 		func(data cmtevents.EventData) {
 			conR.broadcastHasVoteMessage(data.(*types.Vote))
-			conR.updateRoundStateNoCsLock()
+
+			// update reactor's view of round state
+			// NOTE this is safe to do without locking cs because the eventBus is
+			// synchronous. If it were not, we could pass rs in this event
+			// instead
+			conR.rsMtx.Lock()
+			defer conR.rsMtx.Unlock()
+			conR.rs = conR.conS.getRoundState()
 		}); err != nil {
 		conR.Logger.Error("Error adding listener for events (Vote)", "err", err)
 	}
 
 	if err := conR.conS.evsw.AddListenerForEvent(subscriber, types.EventProposalBlockPart,
 		func(data cmtevents.EventData) {
 			conR.broadcastHasProposalBlockPartMessage(data.(*BlockPartMessage))
-			conR.updateRoundStateNoCsLock()
+
+			// update reactor's view of round state
```

Comment:
This is a good question, not sure if I trust the concurrency model of the synchronous bus.